src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ActivitiesService.kt

```diff
@@ -74,9 +76,55 @@ class ActivitiesService(
       )
       log.debug("OffenderProgramProfile created with Nomis id = $offenderProgramReferenceId")
 
-      CreateOffenderProgramProfileResponse(offenderProgramReferenceId)
+      OffenderProgramProfileResponse(offenderProgramReferenceId)
     }
 
+  fun endOffenderProgramProfile(
+    courseActivityId: Long,
+    bookingId: Long,
+    dto: EndOffenderProgramProfileRequest
+  ): OffenderProgramProfileResponse =
+
+    offenderProgramProfileRepository.findByCourseActivityIdAndOffenderBookingIdAndAlloc(courseActivityId, bookingId)
+      ?.run {
+
+        dto.endReason?.also {
+          programServiceEndReasonRepository.findByIdOrNull(ProgramServiceEndReason.pk(it))?.also {
+            this.endReason = it
+          }
+            ?: throw BadDataException("End reason code=${dto.endReason} is invalid")
+        }
+        endDate = dto.endDate
+        endComment = dto.endComment
+
+        telemetryClient.trackEvent(
+          "offender-program-profile-ended",
+          mapOf(
+            "id" to offenderProgramReferenceId.toString(),
+            "courseActivityId" to courseActivity?.courseActivityId.toString(),
+            "bookingId" to offenderBooking.bookingId.toString(),
+          ),
+          null
+        )
+        log.debug("OffenderProgramProfile with Nomis id = $offenderProgramReferenceId set as ended")
+
+        OffenderProgramProfileResponse(offenderProgramReferenceId)
+      }
+
+      ?: throwError(courseActivityId, bookingId)
```

Comment:
I had to create a function here as it didnt like an inline block

Reply:
For an inline block you can do `?: run {`

Reply:
intellij did this when i inlined the fun, but then that had a compilation error. Ive just had another look and the inline block would need to be cast to OffenderProgramProfileResponse which looks messier than a fun i think.

Reply:
Run seems to work on my machine, it seems to understand that the run block doesn't actually return? Also `throwError` is a very generic function name for a very specific error scenario.